Clappr/Classes/Base/BaseObject.swift

```diff
@@ -1,123 +1,79 @@
     }
 
-    open func trigger(_ eventName: String, userInfo: [AnyHashable: Any]?) {
-        notificationCenter().post(name: Notification.Name(rawValue: eventName), object: self, userInfo: userInfo)
-
-        if type(of: self) != BaseObject.self {
-            Logger.logDebug("[\(eventName)] triggered with \(String(describing: userInfo))", scope: logIdentifier())
-        }
+    public func trigger(_ eventName: String, userInfo: EventUserInfo) {
+        eventDispatcher.trigger(eventName, userInfo: userInfo)
+        Logger.logDebug("[\(eventName)] triggered with \(String(describing: userInfo))", scope: logIdentifier())
     }
 
     @discardableResult
-    open func listenTo<T: EventProtocol>(_ contextObject: T, eventName: String, callback: @escaping EventCallback) -> String {
-        return on(eventName, callback: callback, contextObject: contextObject.getEventContextObject())
+    public func listenTo<T: EventProtocol>(_ contextObject: T, eventName: String, callback: @escaping EventCallback) -> String {
+        return eventDispatcher.listenTo(contextObject, eventName: eventName, callback: callback)
     }
 
     @discardableResult
-    open func listenToOnce<T: EventProtocol>(_ contextObject: T, eventName: String, callback: @escaping EventCallback) -> String {
-        return once(eventName, callback: callback, contextObject: contextObject.getEventContextObject())
+    public func listenToOnce<T: EventProtocol>(_ contextObject: T, eventName: String, callback: @escaping EventCallback) -> String {
+        return eventDispatcher.listenToOnce(contextObject, eventName: eventName, callback: callback)
     }
 
-    open func stopListening() {
-        for (_, event) in events {
-            notificationCenter().removeObserver(event.eventHandler)
-        }
-
-        events.removeAll()
+    public func stopListening() {
+        eventDispatcher.stopListening()
     }
 
-    open func stopListening(_ listenId: String) {
-        off(listenId)
+    public func stopListening(_ listenId: String) {
+        eventDispatcher.stopListening(listenId)
     }
 
-    open func getEventContextObject() -> BaseObject {
-        return self
+    public func getEventDispatcher() -> EventDispatcher {
+        return eventDispatcher
     }
 
-    fileprivate func createListenId(_ eventName: String, contextObject: BaseObject) -> String {
-        let contextObjectHash = ObjectIdentifier(contextObject).hashValue
-        return eventName + String(contextObjectHash) + String(Date().timeIntervalSince1970)
+    public func trigger(_ event: Event) {
+        trigger(event.rawValue)
     }
 
-    fileprivate func notificationCenter() -> NotificationCenter {
-        return NotificationCenter.default
+    public func trigger(_ event: Event, userInfo: EventUserInfo) {
+        trigger(event.rawValue, userInfo: userInfo)
     }
 
-    fileprivate func logIdentifier() -> String {
+    func logIdentifier() -> String {
         if let plugin = self as? Plugin {
```

Comment:
What if you check if self conforms with NamedType protocol?

Reply:
Hey @CristianDRM, yes I can do it. But I didn't include the `NamedType` protocol yet. This PR is related only with `BaseOject` refactoring. Do you think I should include `NamedType` protocol now?

Reply:
As you mention, If the NamedType was not implemented yet, it is better to not add this protocol to just use here